llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -169,29 +169,42 @@ void ModuleHashGenerator::calculateGONumAndGOIMap(const Module &M) {
 #endif
 }
 
-void ModuleHashGenerator::updateDigestUseContributions(
-    MD5 &GOHash, const GOVec &Contributions) {
-  for (const GlobalObject *const G : Contributions) {
-    GOHash.update(GOIMap[G].InitialDigest.Bytes);
-  }
+size_t ModuleHashGenerator::updateDigestUseContributions(
+    const GlobalObject *GO, MD5 &GOHash, const GOVec &Contributions,
+    bool UseRepoDefinitionMD) {
   // GOHash/Module is changed if Contributions is not empty.
   Changed = Changed || !Contributions.empty();
+
+  auto LoopPoint = std::numeric_limits<size_t>::max();
+  for (const GlobalObject *const G : Contributions) {
+    // Although 'contributions' and 'dependencies' respective arrows point are
+    // in different direction,  they can affect each other and form loops.
+    // Consider the following example:
+    // int var = 0;
+    // int Func() {
+    //    var = var + 5;
+    //    return var;
+    //  }
```

Comment:
Taking a step back for a moment. The graph generated by your code example is captured like this:

![exa](https://user-images.githubusercontent.com/2496470/89161613-219b4280-d573-11ea-9ae4-fefcd5560ea5.png)

The blue edge labelled “d” indicates a dependency, and the red edge labelled “c” indicates a contribution. (Note that I've renamed both the variable and function according to the LLVM-style.) 

In this graph we encode the relationship between `Var` and `func` such that a modification to either will trigger recompilation of both. This reflects that optimizations to `func` can update the final binary representation of `Var` and vice-versa.

“contributions” were initially known as "reverse dependencies", and that's just what they are: dependencies in the graph but with the direction of the edge reversed because the code that scans the definition of `Var` is unable to see or record the relationship with `func`. This first graph is semantically identical to this one:

![exb](https://user-images.githubusercontent.com/2496470/89162050-c87fde80-d573-11ea-9806-9abba6cd6a7a.png)

Here I've simply swapped the contribution/reverse-dependency for a normal (forward) dependency. The meaning is unchanged.

Is it worth considering a implementing pass which does just this transformation: convert the contributions to dependencies and then just record the resulting graph in the hash?

Reply:
> Is it worth considering a implementing pass which does just this transformation: convert the contributions to dependencies and then just record the resulting graph in the hash?

Good advice. I think it is worth. In additions, it is just a line changed to convert the contributions to the dependencies
Current implementation:
```
  for (auto &GO : Result.Contributions) {
    assert(isa<GlobalVariable>(GO) &&
           "Only global variables can have contributions!");
    assert(isa<llvm::Function>(G) && "All contributions are functions!");
    GOIMap[GO].Contributions.emplace_back(G);                    <---- Only need to change this line;
  }
```
Line:
```
    GOIMap[GO].Contributions.emplace_back(G); 
```

Changed to 
```
    GOIMap[GO].Dependencies.emplace_back(G); 
```

After this change, we could avoid some questions you mentioned below.

Reply:
It's not clear to me how this change results in the direction of the arrow being reversed. Could you explain?

Reply:
Two steps have been invoked when calculating the `Contributions` and `Dependencies` of each global object.

Step 1: Calculate the `Dependencies` and `ContributedToGVs` for all functions and variables in the input module. Please note, a) `Dependencies` and `ContributedToGVs` are in the **reverse** direction. b) The RepoHashCalculator class is used to calculate each global object **independently**.

Step 2: Covert `ContributedToGVs` to `Contributions` for all functions in the input module. Please note, a) Only functions have `ContributedToGVs`. b) `ContributedToGVs` is **reverse dependencies**. c) `Contributions` and `Dependencies` is in the same directory after step 2 (from my understanding).

I will explain these steps using the following example:

```C
int Var = 0;
int func1() {
   Var += 5;
   return Var;
}
int func2(int V) {
   Var += V;
   return Var;
}
```

**_Step 1:_** 
Dump the `Dependencies` and `ContributedToGVs` using the option `-debug-only=repo-hash-calculator`.
```
$ clang -S -emit-llvm --target=x86_64-pc-linux-gnu-repo -mllvm -debug-only=repo-hash-calculator test.c -o repo.o
```
You could see the following debug information:
```
GO name is: func1. Its Dependencies are: Var. Its ContributedToGVs are: Var.
GO name is: func2. Its Dependencies are: Var. Its ContributedToGVs are: Var.
```
We can see:
 1) `Var` doesn't have `Dependencies` and `ContributedToGVs`.
 2) Both `func1` and `func2` have a dependency `Var` and a `ContributedToGVs` `Var`.
    
    Like the graph in your comment:
	![exa](https://user-images.githubusercontent.com/2496470/89161613-219b4280-d573-11ea-9ae4-fefcd5560ea5.png)
	
	The blue edge labelled “d” indicates a dependency, and the red edge labelled “c” indicates a `ContributedToGVs`.
	
In summary, the following information is collected after step 1:
```
a) ContributedToGVs[`Var`] = []
   Dependencies[`Var`] = []
b) ContributedToGVs[`func1`] = [`Var`]
   Dependencies[`func1`] = [`Var`]
c) ContributedToGVs[`func2`] = [`Var`]
   Dependencies[`func2`] = [`Var`]
```

**_Step 2:_** 

Dump the `Dependencies` and `Contributions` using the option `-debug-only=prepo-digest`.
```
$ clang -S -emit-llvm --target=x86_64-pc-linux-gnu-repo -mllvm -debug-only=prepo-digest test.c -o repo.o
```
You could see the following debug information:
```
GO Name:func1
GOInfo:
        Initial Digest: 7dfb010ea6cd8aecb8bf10886b3cc631
        Contributions: [ ]
        Dependencies: [ Var]

GO Name:Var
GOInfo:
        Initial Digest: 3eb6bded68925949c48b7bc8959c8dbe
        Contributions: [ func1,func2]
        Dependencies: [ ]

GO Name:func2
GOInfo:
        Initial Digest: 82901993642a7935214b498e69e8cdbe
        Contributions: [ ]
        Dependencies: [ Var]
```

In this step, the GO's `Dependencies` will keep the same as step1. The `Contributions` initial value is empty. The `Contributions` is collected for all variables by going through all functions.

1) Convert `func1`'s `ContributedToGVs` to `Var`'s `Contributions`.
   This step loops all global variables in the `ContributedToGVs` of `func1`(only `Var` in this example) and add the function `func1` into the `Contributions` of `Var`.
   After this step, the `Dependencies` and `Contributions` for all global objects:
```
a) Contributions[`Var`] = [`func1`]
   Dependencies[`Var`] = []
b) Contributions[`func1`] = []
   Dependencies[`func1`] = [`Var`]
c) Contributions[`func2`] = []
   Dependencies[`func2`] = [`Var`]
```
2) Convert `func2`'s `ContributedToGVs` to `Var`'s `Contributions`.
   Similar to `func1`, the `func2` ia added into the `Contributions` of `Var`.
   After this step, the `Dependencies` and `Contributions` for all global objects:
```
a) Contributions[`Var`] = [`func1`, `func2`]
   Dependencies[`Var`] = []
b) Contributions[`func1`] = []
   Dependencies[`func1`] = [`Var`]
c) Contributions[`func2`] = []
   Dependencies[`func2`] = [`Var`]
```

If we removed the `Contributions` structure and covert `ContributedToGVs` to `Dependencies` instead in step 2, we will get a single resulting graph as you advised.

![exb](https://user-images.githubusercontent.com/2496470/89162050-c87fde80-d573-11ea-9806-9abba6cd6a7a.png)

Reply:
> Step 1: Calculate the `Dependencies` and `ContributedToGVs` for all functions and variables in the input module. Please note, a) `Dependencies` and `ContributedToGVs` are in the **reverse** direction. 

Both of them? Your later text and examples appear to contradict this since they show the _dependency_ arrows in the expected direction. e.g.:

> ```c
> int Var = 0;
> int func(void) {
>    Var += 5;
>    return Var;
> }
> ```
>
> ```clang -S -emit-llvm --target=x86_64-pc-linux-gnu-repo -mllvm -debug-only=repo-hash-calculator test.c -o repo.o```
> You could see the following debug information:
> ```GO name is: func. Its Dependencies are: Var. Its ContributedToGVs are: Var.```

(Note that I removed `func2` from the example: I’m not sure that it adds to the discussion.) That gives us the graph in the first diagram above and is the same as my previous example except that the edge labelled “c” is what you’re calling “ContributedToGVs” and I was calling “contributions”.

> Step 2: Covert `ContributedToGVs` to `Contributions` for all functions in the input module. 

This is a step that I didn’t appreciate/had forgotten and might explain some of my difficulty in understanding this PR.

> If we removed the `Contributions` structure and covert `ContributedToGVs` to `Dependencies` instead in step 2, we will get a single resulting graph as you advised.

Reply:
> Both of them? Your later text and examples appear to contradict this since they show the dependency arrows in the expected direction.
>
Thanks for correcting my sentence.

I will rename `ContributedToGVs`  to `Contributions` in the step 1. In step 2, I will covert `Contributions` to `Dependencies` to get a single resulting graph. Are you ok with the changes?

Reply:
> I will rename ContributedToGVs to Contributions in the step 1. In step 2, I will covert Contributions to Dependencies to get a single resulting graph. Are you ok with the changes?

Yes, I think this makes sense.

Reply:
This sounds good. If from some point we can just deal with dependencies rather than both dependencies and contributions then that sounds like it will make things quite a bit clearer.

Reply:
I have removed the contributions and only dealt with the dependencies in the hash calculation.